sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,451 @@
+  private def acquire(): Unit = acquireLock.synchronized {
+    val newAcquiredThreadInfo = AcquiredThreadInfo()
+    val waitStartTime = System.currentTimeMillis
+    def timeWaitedMs = System.currentTimeMillis - waitStartTime
+    def isAcquiredByDifferentThread = acquiredThreadInfo != null &&
+      acquiredThreadInfo.threadRef.get.isDefined &&
+      newAcquiredThreadInfo.threadRef.get.get.getId != acquiredThreadInfo.threadRef.get.get.getId
+
+    while (isAcquiredByDifferentThread && timeWaitedMs < conf.lockAcquireTimeoutMs) {
+      acquireLock.wait(10)
+    }
+    if (isAcquiredByDifferentThread) {
+      val stackTraceOutput = acquiredThreadInfo.threadRef.get.get.getStackTrace.mkString("\n")
+      val msg = s"RocksDB instance could not be acquired by $newAcquiredThreadInfo as it " +
+        s"was not released by $acquiredThreadInfo after $timeWaitedMs ms.\n" +
+        s"Thread holding the lock has trace: $stackTraceOutput"
+      logError(msg)
+      throw new IllegalStateException(s"$loggingId: $msg")
+    } else {
+      acquiredThreadInfo = newAcquiredThreadInfo
+      // Add a listener to always release the lock when the task (if active) completes
+      Option(TaskContext.get).foreach(_.addTaskCompletionListener[Unit] { _ => this.release() })
+      logInfo(s"RocksDB instance was acquired by $acquiredThreadInfo")
+    }
+  }
+
+  private def release(): Unit = acquireLock.synchronized {
+    acquiredThreadInfo = null
+    acquireLock.notifyAll()
+  }
+
+  private def openDB(): Unit = {
+    assert(db == null)
+    db = NativeRocksDB.open(dbOptions, workingDir.toString)
+    logInfo(s"Opened DB with conf ${conf}")
+  }
+
+  private def closeDB(): Unit = {
+    if (db != null) {
+      db.close()
+      db = null
+    }
+  }
+
+  /** Create a native RocksDB logger that forwards native logs to log4j with correct log levels. */
+  private def createLogger(): Logger = {
+    val dbLogger = new Logger(dbOptions) {
+      override def log(infoLogLevel: InfoLogLevel, logMsg: String) = {
+        // Map DB log level to log4j levels
+        // Warn is mapped to info because RocksDB warn is too verbose
+        // (e.g. dumps non-warning stuff like stats)
+        val loggingFunc: ( => String) => Unit = infoLogLevel match {
+          case InfoLogLevel.FATAL_LEVEL | InfoLogLevel.ERROR_LEVEL => logError(_)
+          case InfoLogLevel.WARN_LEVEL | InfoLogLevel.INFO_LEVEL => logInfo(_)
+          case InfoLogLevel.DEBUG_LEVEL => logDebug(_)
+          case _ => logTrace(_)
+        }
+        loggingFunc(s"[Native-${infoLogLevel.getValue}] $logMsg")
+      }
+    }
+
+    var dbLogLevel = InfoLogLevel.ERROR_LEVEL
+    if (log.isWarnEnabled) dbLogLevel = InfoLogLevel.WARN_LEVEL
+    if (log.isInfoEnabled) dbLogLevel = InfoLogLevel.INFO_LEVEL
+    if (log.isDebugEnabled) dbLogLevel = InfoLogLevel.DEBUG_LEVEL
+    dbOptions.setLogger(dbLogger)
+    dbOptions.setInfoLogLevel(dbLogLevel)
+    logInfo(s"Set DB native logging level to $dbLogLevel")
```

Comment:
We can be more specified. `DB` -> `RocksDB`

Reply:
Sure, done in the next commit.